Anisha Gharat

BE Comps Batch E

Roll no.: 68

UID: 2018230071

# **EXPERIMENT NO. 2B**

**Aim:** To apply AHP method to "Machinability Evaluation of  Work Materials" as given in the book.

# **TextBook Sum**

In [1]:
import pandas as pd

tf = pd.DataFrame(columns = ["VC", "CF", "PI"])
tf["VC"] = [710,900,1630,1720,120,160]
tf["CF"] = [400,415,440,235,1150,1750]
tf["PI"] = [28,38,59,43,8,19]
print("Book Sum Dataset:")
tf

Book Sum Dataset:


VC    CF  PI
0   710   400  28
1   900   415  38
2  1630   440  59
3  1720   235  43
4   120  1150   8
5   160  1750  19

In [2]:
nf=pd.DataFrame(columns=["VC", "CF", "PI"])
nf["VC"] = [0.4128, 0.5233, 0.9477, 1, 0.0698, 0.0932]
nf["CF"] = [0.5875, 0.5663, 0.5341, 1, 0.2043, 0.1343]
nf["PI"] = [0.2857, 0.2105, 0.1356, 0.1860, 1, 0.4211]
print("Normalised data of the attributes:")
nf

Normalised data of the attributes:


VC      CF      PI
0  0.4128  0.5875  0.2857
1  0.5233  0.5663  0.2105
2  0.9477  0.5341  0.1356
3  1.0000  1.0000  0.1860
4  0.0698  0.2043  1.0000
5  0.0932  0.1343  0.4211

In [3]:
ef=pd.DataFrame(columns=["VC", "CF", "PI"])
ef["VC"] = [1,0.255,0.255]
ef["CF"] = [0.745,1,0.5]
ef["PI"] = [0.745,0.5,1]
print("Normalised Pairwise comparison matrix:")
ef

Normalised Pairwise comparison matrix:


VC     CF     PI
0  1.000  0.745  0.745
1  0.255  1.000  0.500
2  0.255  0.500  1.000

In [4]:
ef['Geometric Mean'] = ef.prod(axis=1)**(1/3)
ef

VC     CF     PI  Geometric Mean
0  1.000  0.745  0.745        0.821809
1  0.255  1.000  0.500        0.503311
2  0.255  0.500  1.000        0.503311

In [5]:
gmsum = ef['Geometric Mean'].sum()
print("Sum of Geometric mean: ",gmsum)

Sum of Geometric mean:  1.828431625150487


In [6]:
ef['Normalized Geometric Mean'] = ef['Geometric Mean']/gmsum
ef

VC     CF     PI  Geometric Mean  Normalized Geometric Mean
0  1.000  0.745  0.745        0.821809                   0.449461
1  0.255  1.000  0.500        0.503311                   0.275269
2  0.255  0.500  1.000        0.503311                   0.275269

In [7]:
ngmsum=ef['Normalized Geometric Mean'].sum()
print("Normalized Geometric Mean sum: ",ngmsum)

Normalized Geometric Mean sum:  0.9999999999999998


In [8]:
del ef['Geometric Mean']
ef

VC     CF     PI  Normalized Geometric Mean
0  1.000  0.745  0.745                   0.449461
1  0.255  1.000  0.500                   0.275269
2  0.255  0.500  1.000                   0.275269

In [9]:
af = pd.DataFrame()
af = ef['Normalized Geometric Mean']
af

0    0.449461
1    0.275269
2    0.275269
Name: Normalized Geometric Mean, dtype: float64

In [10]:
bf=ef['VC']
cf=ef['CF']
df=ef['PI']

In [11]:
vc=bf*af[0]
cf=cf*af[1]
pi=df*af[2]
vcd = pd.DataFrame(vc)
cfd = pd.DataFrame(cf)
pid = pd.DataFrame(pi)

In [12]:
ff = pd.concat([vcd, cfd, pid], axis=1)
print("Check for Consistency")
ff

Check for Consistency


VC        CF        PI
0  0.449461  0.205076  0.205076
1  0.114613  0.275269  0.137635
2  0.114613  0.137635  0.275269

In [13]:
A3=pd.DataFrame(columns=['A3=A1*A2'])
A3['A3=A1*A2']=ff.sum(axis=1)
A3

A3=A1*A2
0  0.859613
1  0.527517
2  0.527517

In [14]:
a3=A3['A3=A1*A2']
a4=a3/af
A4=pd.DataFrame(columns=['A4'])
A4['A4']=a4
A4

A4
0  1.912540
1  1.916365
2  1.916365

In [15]:
Ymax=A4['A4'].mean()
print("Average of A4 (Ymax): ",Ymax)

Average of A4 (Ymax):  1.9150902212482916


In [16]:
n=3
CI = (Ymax-n)/(n-1)
print("Consistency Index (CI) = ",CI)
RI = 0.52
CR = CI/RI
print("Consistency Ratio (CR) = ",CR)

Consistency Index (CI) =  -0.5424548893758542
Consistency Ratio (CR) =  -1.0431824795689504


**CR = -1.0431824795689504 is less than 0.10. Hence the matrix is consistent.**

In [17]:
jf = pd.DataFrame(columns=['Normalized Geometric Mean','Percentage of Weightage'])
jf['Normalized Geometric Mean'] = af
jf['Percentage of Weightage'] = af*100
print("Model Synthesis")
jf

Model Synthesis


Normalized Geometric Mean  Percentage of Weightage
0                   0.449461                44.946111
1                   0.275269                27.526944
2                   0.275269                27.526944

**VC is a beneficial attribute and CF & PI are non-beneficial attributes**

In [18]:
tf

VC    CF  PI
0   710   400  28
1   900   415  38
2  1630   440  59
3  1720   235  43
4   120  1150   8
5   160  1750  19

In [19]:
vcmax = int(tf["VC"].max())
cfmin = int(tf["CF"].min())
pimin = int(tf["PI"].min())
print("Min-Max values:")
print("Max of VC: ",vcmax)
print("Min of CF: ",cfmin)
print("Min of PI: ",pimin)

Min-Max values:
Max of VC:  1720
Min of CF:  235
Min of PI:  8


In [20]:
w1=0.7142
w2=0.1429
w3=0.1429

In [21]:
ndf = tf
ndf["VC"] = ndf["VC"]/vcmax
ndf["CF"] = cfmin/ndf["CF"]
ndf["PI"] = pimin/ndf["PI"]
print("Normalized Decision matrix:")
ndf

Normalized Decision matrix:


VC        CF        PI
0  0.412791  0.587500  0.285714
1  0.523256  0.566265  0.210526
2  0.947674  0.534091  0.135593
3  1.000000  1.000000  0.186047
4  0.069767  0.204348  1.000000
5  0.093023  0.134286  0.421053

In [22]:
ndf["VC"] = ndf["VC"]*w1
ndf["CF"] = ndf["CF"]*w2
ndf["PI"] = ndf["PI"]*w3
ndf["Ai"] = ndf.sum(axis=1)
ndf['Rank'] = ndf['Ai'].rank(ascending = 0) 
print("Calculation of Ai and Ranking of Alternatives:")
ndf

Calculation of Ai and Ranking of Alternatives:


VC        CF        PI        Ai  Rank
0  0.294815  0.083954  0.040829  0.419597   4.0
1  0.373709  0.080919  0.030084  0.484713   3.0
2  0.676829  0.076322  0.019376  0.772527   2.0
3  0.714200  0.142900  0.026586  0.883686   1.0
4  0.049828  0.029201  0.142900  0.221929   5.0
5  0.066437  0.019189  0.060168  0.145795   6.0

In [23]:
ndf = ndf.set_index('Rank')
ndf = ndf.sort_index() 
print("Rank wise Sorting:")
ndf

Rank wise Sorting:


VC        CF        PI        Ai
Rank                                        
1.0   0.714200  0.142900  0.026586  0.883686
2.0   0.676829  0.076322  0.019376  0.772527
3.0   0.373709  0.080919  0.030084  0.484713
4.0   0.294815  0.083954  0.040829  0.419597
5.0   0.049828  0.029201  0.142900  0.221929
6.0   0.066437  0.019189  0.060168  0.145795

**Hence W4 is the best choice.**